cmds/webboot/webboot.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/u-root/webboot/pkg/menu"
+)
+
+const (
+	downloadByLinkLabel = "Download by link"
+	tcURL               = "http://tinycorelinux.net/10.x/x86_64/release/TinyCorePure64-10.1.iso"
+)
+
+// bookmark record the list of iso webboot allow user to download
+var bookmark = map[string]DownloadedDistro{
+	"tinycore": DownloadedDistro{
+		isDefault: false,
+		url:       tcURL,
+		label:     "Download Tinycore v10.1",
+		isoName:   "TinyCorePure64-10.1.iso",
+		group:     "Tinycore",
+	},
+}
+
+// BootMedia contains required information of a bootable iso
+// no matter it is cached or downloaded
+type BootMedia interface {
+	Path() string
+	Name() string
+	// todo: may have Kernel(), Config() etc later
+}
+
+// DownloadedDistro contains information of the iso user want to download
+type DownloadedDistro struct {
+	isDefault bool
+	label     string
+	isoName   string
+	group     string
+	url       string
+	path      string
+}
+
+// CachedDistro contains information of the iso cached in the memory
+type CachedDistro struct {
+	isDefault bool
+	label     string
+	group     string
+	path      string
+	isoName   string
+}
+
+// GoBack is the backward option
+type GoBack struct {
+	isGoBack bool
+}
+
+// IsDefault is for mark whether this iso is a default choice.
+func (d *DownloadedDistro) IsDefault() bool {
+	return d.isDefault
+}
+
+// Label is the string this iso displays in the menu page.
+func (d *DownloadedDistro) Label() string {
+	return d.label
+}
+
+// Exec performs the following process after the entry is chosen
+func (d *DownloadedDistro) Exec(uiEvents <-chan ui.Event) error {
+	if d.Label() == downloadByLinkLabel {
+		noConstraint := func(input string) (string, string, bool) {
+			return input, "", true
+		}
+		link, err := menu.NewInputWindow("Please input the link", noConstraint, uiEvents)
+		if err != nil {
+			return err
+		}
+		spaceConstraint := func(input string) (string, string, bool) {
+			re := regexp.MustCompile(`[\w]+.iso`)
+			if re.Match([]byte(input)) {
+				return input, "", true
+			}
+			return "", "File name should only contain [a-zA-Z0-9_], and should be end as .iso", false
+		}
+		filename, err := menu.NewInputWindow("Please input the name of iso", spaceConstraint, uiEvents)
```

Comment:
As much as its nice to have polite messages, I think it might be better to have short ones :) Something like "Input ISO name:" would work as well!